[Step 4] JM

alwaysblu opened the pull request:
@AppleCEO 
<br>

안녕하세요 도미닉 step4 Pull Request 드립니다! 😄 

<br>

![step4_01](https://user-images.githubusercontent.com/75533266/122244817-243f2600-cf00-11eb-86fb-40e68ee28454.gif)
![step4_02](https://user-images.githubusercontent.com/75533266/122244835-2903da00-cf00-11eb-9e63-c8cebacf0ef7.gif)
![step4_03](https://user-images.githubusercontent.com/75533266/122244836-2a350700-cf00-11eb-8541-87ace64c3797.gif)
![step4_04](https://user-images.githubusercontent.com/75533266/122244841-2acd9d80-cf00-11eb-8a2a-62ee102a682d.gif)

<br>

# 📌 코드를 작성할 때 고민되었던 점 :

TextView에서 실시간으로 Title 스타일과 Body 스타일을 적용시키는 것에 대해서 고민하였습니다.

<br>

# 📌 조언을 얻고 싶은 부분 :

1. 유닛테스트를 작성하기 위해  Testable한 코드로 리팩토링을 하려고 계획하고 있는데      
Testable한 코드로 리팩토링을 하기위해서 현재 어떤 코드를 어떻게 변화를 주는 것이 좋겠는지에 대한 조언을 얻고 싶습니다!        
(현재 싱글톤과 뷰컨트롤러 프로퍼티를 이용하여 데이터를 이동시키는 부분 등과 같은        
커플링되어 있는 부분들을 풀어주려고 계획하고 있습니다.     
추가적으로 어떤 부분을 리팩터링하면 좋을지 궁금합니다! )

2. step 별 요구사항으로 유닛테스트를 작성하는 기준을 세워봤는데 적절한 기준인지        
 아니면 추가적으로 더 유닛테스트를 작성해야할지에 대한 조언을 얻고 싶습니다!

3. 전반적인 코드에 대한 조언을 얻고 싶습니다!

감사합니다! 😄

alwaysblu committed:
Feat: text style 변경

alwaysblu committed:
Refactor: 실시간으로 폰트의 스타일을 지정하도록함

alwaysblu committed:
Feat: dark mode 추가함

alwaysblu committed:
Feat: searchBar UI 추가

alwaysblu committed:
Feat: 플레이스 홀더 추가함

alwaysblu committed:
Fix: 커서문제 발생

alwaysblu committed:
Fix: 텍스트 뷰 커서 움직이는 문제 해결

alwaysblu committed:
Fex : 스크롤 문제 해결
memoTextView.contentOffset = CGPoint(x: 0,y: 0) 추가함

alwaysblu committed:
Feat: 검색 기능 추가함

alwaysblu committed:
Feat: 삭제 기능 추가함

alwaysblu committed:
Fix: ipad에서 actionSheet나 activityController를 띄울때 위치를 지정안해주면 나는 에러 수정함

alwaysblu committed:
Refactor: 공통함수 extension으로 묶음

alwaysblu committed:
Chore: 공백 제거

alwaysblu committed:
Feat: delete 시 searchedMemoResults 업데이트 기능 추가함

alwaysblu committed:
Feat: searchedMemoResults 업데이트 추가함

alwaysblu committed:
Refactor: locate함수들을 하나의 함수로 리팩터링함

AppleCEO commented on CloudNotes/CloudNotes/Controllers/DetailMemoViewController.swift:
```diff
@@ -14,11 +13,14 @@ class DetailMemoViewController: UIViewController {
+    var memoBody: String?
+    var allText: String?
     var memoListViewController: MemoListViewController?
     
-    lazy var rightNvigationItem: UIButton = {
+    lazy var rightNavigationItem: UIButton = {
```
오타를 수정한 부분 좋습니다! 👍

AppleCEO commented on CloudNotes/CloudNotes/Controllers/DetailMemoViewController.swift:
```diff
@@ -125,10 +160,39 @@ extension DetailMemoViewController: UITextViewDelegate {
-    func filterTitleAndBody(separatedTextArray: [String], completion: @escaping (String, String) -> ()) {
+    private func filterTitleAndBody(separatedTextArray: [String], completion: @escaping (String, String) -> ()) {
         var separatedTextArray = separatedTextArray
         guard separatedTextArray.count > 0, let title = separatedTextArray.first(where: { $0 != "" }), let titleIndex = separatedTextArray.firstIndex(of: title) else {
```
이 줄 가독성이 많이 떨어지네요.
어떻게 개선해 볼 수 있을까요?

alwaysblu replied:
가독성을 높이기 위해 조건과 값을 찾는 부분을 나누어봤는데 적절한 수정인지 혹시 더 나은 방법이 있을지 조언을 얻고 싶습니다!

```swift
guard separatedTextArray.count > 0 else {
    completion("", "")
    return
}
guard let title = separatedTextArray.first(where: { $0 != "" }), let titleIndex = separatedTextArray.firstIndex(of: title) else {
    completion("", "")
    return
}
```

AppleCEO commented on CloudNotes/CloudNotes/Controllers/DetailMemoViewController.swift:
```diff
@@ -143,13 +207,31 @@ extension DetailMemoViewController: UITextViewDelegate {
 
+extension UIViewController {
+    func setDynamicTintColor(dark: UIColor, light: UIColor, traiteCollection: UITraitCollection) -> UIColor {
+        if traiteCollection.userInterfaceStyle == .dark {
+            return dark
+        } else {
+            return light
+        }
```
다크모드를 대응해보셨군요! 💯

AppleCEO commented on CloudNotes/CloudNotes/Controllers/DetailMemoViewController.swift:
```diff
@@ -143,13 +207,31 @@ extension DetailMemoViewController: UITextViewDelegate {
+
+    func locateController(controller: UIViewController) {
+        if UIDevice.current.userInterfaceIdiom == .pad {
+            if let popoverController = controller.popoverPresentationController {
+                popoverController.sourceView = self.view
+                popoverController.sourceRect = CGRect(x: self.view.bounds.midX, y: self.view.bounds.midY, width: 0, height: 0)
+            }
```
이 부분은 if let도 좋네요! 위에 if문과 합치는 방법은 어떨까요?

alwaysblu replied:
합치는 것이 깊이를 낮출 수 있는 좋은 방법인 것 같습니다 😄 

```swift
if UIDevice.current.userInterfaceIdiom == .pad, let popoverController = controller.popoverPresentationController {
    popoverController.sourceView = self.view
    popoverController.sourceRect = CGRect(x: self.view.bounds.midX, y: self.view.bounds.midY, width: 0, height: 0)
}
```

위와 같이 수정해보았습니다!

AppleCEO commented on CloudNotes/CloudNotes/Controllers/MemoListViewController.swift:
```diff
@@ -39,11 +52,21 @@ class MemoListViewController: UIViewController {
     }
     
+    private func setSearchBarLayout() {
+        let safeArea = self.view.safeAreaLayoutGuide
```
이렇게 할당을 한 부분도 참 좋네요~ 🥇

AppleCEO commented on CloudNotes/CloudNotes/Controllers/MemoListViewController.swift:
```diff
@@ -55,34 +78,53 @@ class MemoListViewController: UIViewController {
             return UITableViewCell()
         }
-        cell.configure(with: MemoCache.shared.memoData[indexPath.row])
+        if isThereTextInSearchBar {
+            cell.configure(with: MemoCache.shared.searchedMemoResults[indexPath.row])
+            return cell
+        }
+        cell.configure(with: MemoCache.shared.memoDataList[indexPath.row])
         return cell
```
else 구문을 사용하지 않은 부분도 참 좋네요. 💯

AppleCEO commented on CloudNotes/CloudNotes/Controllers/MemoListViewController.swift:
```diff
@@ -111,3 +156,19 @@ extension MemoListViewController: UITableViewDelegate {
 
+extension MemoListViewController: UISearchBarDelegate {
+    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+        guard searchText != "" else {
+            isThereTextInSearchBar = false
+            self.tableView.reloadData()
+            return
+        }
```
부정의 부정이라서 이해하기 힘든 부분인 것 같네요.
어떻게 개선해 볼 수 있을까요?

alwaysblu replied:
부정의 부정은 긍정이므로 if로 변경해보았습니다!

```
if searchText == "" {
    isThereTextInSearchBar = false
    self.tableView.reloadData()
    return
}
```

AppleCEO commented on CloudNotes/CloudNotes/Controllers/MemoListViewController.swift:
```diff
@@ -111,3 +156,19 @@ extension MemoListViewController: UITableViewDelegate {
+    }
+    
+    private func updateSearchedMemoResult(searchText: String) {
+        MemoCache.shared.searchedMemoResults = MemoCache.shared.memoDataList.filter({($0.allText?.contains(searchText) ?? false)})
```
조금 더 익숙해질 때까지 기본형을 쓰는 것이 좋겠습니다. 또한 너무 길어서 보기 힘든 점이 있네요!

AppleCEO commented on CloudNotes/CloudNotes/Views/Cell/MemoListTableViewCell.swift:
```diff
@@ -56,17 +56,19 @@ class MemoListTableViewCell: UITableViewCell {
         NSLayoutConstraint.activate([
             lastModifiedDate.topAnchor.constraint(equalTo: title.bottomAnchor, constant: 5),
             lastModifiedDate.leadingAnchor.constraint(equalTo: layoutGuide.leadingAnchor, constant: 10),
-            lastModifiedDate.trailingAnchor.constraint(equalTo: body.leadingAnchor, constant: -50),
+            lastModifiedDate.trailingAnchor.constraint(equalTo: body.leadingAnchor, constant: 0),
```
-50이 0으로 된 이유는 무엇인가요?

alwaysblu replied:
lastModifiedDate Label과 Preview Label의 관계가 유동적인 관계가 아니라

Preview Label의 leading을 safeArea의 leading과 150 만큼 떨어지도록 고정해 놓았기 때문에 

-50은 의미가 없다고 생각하여 0으로 변경하였습니다!

AppleCEO commented on CloudNotes/CloudNotes/Controllers/DetailMemoViewController.swift:
```diff
@@ -43,11 +49,12 @@ class DetailMemoViewController: UIViewController {
             let actionCancel = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
             alert.addAction(actionCancel)
         }
+        locateController(controller: alert)
         self.present(alert, animated: true, completion: nil)
     }
     
-    func deleteMemo(indexPath: IndexPath) {
-        CoreData.shared.deleteMemoListItem(item: MemoCache.shared.memoData[indexPath.row])
+    func deleteMemo(indexPathRow: Int) {
```
indexPath를 Int형으로 바꾸면서 어떤 점이 좋아지고 어떤 부분이 안 좋아졌을까요?

alwaysblu replied:
indexPath를 Int형으로 바꾸면서 좋아진 점은 

MemoCache.shared.memoDataList.firstIndex(of: MemoCache.shared.searchedMemoResults[indexPath.row]) 로 부터

얻은 인덱스 값을 직접 인자에 대입할 수 있어졌다는 점이 있습니다!

단점으로는 indexPath의 section이 여러개가 되었을 때 문제가 발생할 수 있다는 점이 있습니다!

section이 고려되었을 때 문제가 되지 않는 방법을 생각해보겠습니다!

AppleCEO commented on CloudNotes/CloudNotes/Views/Cell/MemoListTableViewCell.swift:
```diff
@@ -56,17 +56,19 @@ class MemoListTableViewCell: UITableViewCell {
             lastModifiedDate.leadingAnchor.constraint(equalTo: layoutGuide.leadingAnchor, constant: 10),
-            lastModifiedDate.trailingAnchor.constraint(equalTo: body.leadingAnchor, constant: -50),
+            lastModifiedDate.trailingAnchor.constraint(equalTo: body.leadingAnchor, constant: 0),
             lastModifiedDate.bottomAnchor.constraint(equalTo: layoutGuide.bottomAnchor, constant: -5),
```
이런 -5, -50, 5, 10 등의 값을 관리하기에 좋은 방법이 있을까요? 
어떤 이유로 이 값이 들어갔는지 지금은 기억하더라도 나중에는 잊어버릴 수도 있겠네요!

alwaysblu replied:
주석을 남겨주거나 상수 테이블에 의미 있는 변수명으로 관리하는 방법이 있을 것 같습니다!

AppleCEO reviewed (commented):
안녕하세요 도미닉입니다.
JM 고생하셨습니다!

유닛테스트를 작성하기 위해 Testable한 코드로 리팩토링을 하려고 계획하고 있는데
Testable한 코드로 리팩토링을 하기위해서 현재 어떤 코드를 어떻게 변화를 주는 것이 좋겠는지에 대한 조언을 얻고 싶습니다!
(현재 싱글톤과 뷰컨트롤러 프로퍼티를 이용하여 데이터를 이동시키는 부분 등과 같은
커플링되어 있는 부분들을 풀어주려고 계획하고 있습니다.
추가적으로 어떤 부분을 리팩터링하면 좋을지 궁금합니다! )
- 잘나뉘어져 있어서 테스트하기 어려운 구조는 아니라고 생각합니다. 어떤 부분을 고민하셨고 어느 부분이 테스트하기 어렵다고 생각하는지 이야기를 해주시면 같이 고민해 볼 수 있겠습니다~

step 별 요구사항으로 유닛테스트를 작성하는 기준을 세워봤는데 적절한 기준인지
아니면 추가적으로 더 유닛테스트를 작성해야할지에 대한 조언을 얻고 싶습니다!
- CloudNotesTests 파일에 테스트가 비어있는데 테스트 코드를 작성하셨나요?

전반적인 코드에 대한 조언을 얻고 싶습니다!

![image](https://user-images.githubusercontent.com/47517256/122510301-59eb2880-d040-11eb-9ace-1418a0a82869.png)

아이폰에서 제목의 글자 하나를 지우려고 하니 Thread 1: "NSMutableRLEArray 
objectAtIndex:effectiveRange:: Out of bounds" 에러가 발생하네요! 이 부분 수정해보세요~

alwaysblu committed:
Fix: 기존 메모 수정시 발생하는 오류 해결
self?.updateProperties(title: title, body: body, allText: allText, indexPath: indexPath) 를 추가해서 프로퍼티를 업데이트 시켜줌

alwaysblu committed:
Refactor: if문과 합침

alwaysblu committed:
Refactor: 부정의 부정을 긍정으로 변경함

alwaysblu committed:
Refactor: 레이아웃 주석 추가함

alwaysblu commented:
@AppleCEO 

코드리뷰 정말 감사드립니다! 😄 

도미닉께서 주신 피드백에 맞게 코드를 수정해 보았습니다.

알맞게 코드를 수정한 것인지 조언을 얻고 싶습니다!

일단 step 별 요구사항들 부터 유닛테스트를 작성해보려고 하는데 괜찮은 선택일지에 대한 조언을 얻고 싶습니다! 

감사합니다!!

alwaysblu committed:
Remove: 안쓰는 파일 제거함

AppleCEO commented:
그동안 고생 많으셨습니다~ DM으로 머지 요청해주셔서 머지드립니다. 다른 내용도 편하게 질문 주시면 언제든 답변 드릴게요!
